(wip) Improve caching for streams.

showell opened the pull request:
This series has two major goals:
* Add a super-slim, super-fast cache for active user ids.
* Streamline our stream cache not to have realm data.

This series is basically structured so that the first 16 commits should be fairly uncontroversial, but the last five are going to require more scrutiny, and the very last commit might be rejected.

The commits here may change a bit, but here's an overview of the progression.

Add test coverage:

~~~
pick bebe8e0 tests: Count queries for home page.
pick f49bde8 Add test_num_queries_with_streams() for home page.
~~~

Do some basic code cleanup, partly in support of the upcoming commit, but justifiable in their own right:

~~~
pick e8e60a4 Simplify get_stream_cache_key().
pick c3bed5c Only require realm_id for get_active_user_dicts_in_realm().
pick 858e06c Only require realm_id for active_user_ids().
~~~

**Add the user_ids cache**

~~~
pick fdd7059 Cache active_user_ids() more directly.
~~~

Do some simple prep commits that stand on their own merits:

~~~
pick 8a9b1e9 Use `prereg.users.set()` in do_invite_users.
pick c2825a6 Extract stream_name_in_use().
pick aed8d50 Have exclude_topic_mutes() accept a stream id.
pick df94e57 Use update_fields in do_deactivate_stream.
pick 83efcae Only require ids for finding DefaultStream objects.
~~~

Try to streamline our Zephyr codes.  I think these stand on their own merit, but they're also in support of streamlining our cache.

~~~
pick d0fe4a3 Extract is_zephyr_realm local var for bulk operations.
pick 4a8ae80 Add realm param to Stream.is_public().
~~~

Do some simple prep commits that mostly stand on their own merits:

~~~
pick 70b46be Only require stream_id in subscribed_to_stream().
pick a575820 Only require stream_id in  private_stream_user_ids().
~~~

**Make the stream cache slimmer**

~~~
pick 8175a3c Make the stream cache slimmer.
~~~

If we can get everything merged up to the above commit, I'd be pretty happy.  The subsequent commits have a less clear risk/reward tradeoff.  One nice thing about the following commits is that they helped prove to me that the prior commit is valid, and we don't lose anything by removing clunky out-of-date realm data from our stream cache.

The next couple commits are fine, but they're not a major win unless you go to the end.

~~~
pick 431bfad Use stream ids in various tests.
pick 6b609fd Extract Stream.num_subscribers_for_stream_id.
~~~

This next commit is superficially ugly, and it's driven by the end commit:

~~~
pick 1caf1d2 Avoid shadowing stream variables.
~~~

The following commit, while driven by the refactoring, makes sense to me on its own merits, but it deserves some scrutiny:

~~~
pick 63b530c Clear caches more aggressively in do_rename_stream().
~~~

Finally, there's this last commit, which I suspect you may reject, but I encourage you to read its long commit message.
~~~
pick ece529a Streamline stream objects that we get out of cache.
~~~

timabbott commented:
I merged the first batch of commits, since they seemed like clear improvements.  The first thing that worries me is the change to make `Stream.is_public()` require passing in a realm object; that just feels really ugly as an interface.  I think we should consider instead doing some sort of denormalization there; we've been meaning to make "public history" and "anyone can join" separate settings, and maybe we should just do that in the database right now (even if we don't add separate UI).

For the StreamLite code, I'm not super convinced at present; but I think we might want to look into whether there's a way we could tag a Django Stream object as not allowing evaluation (e.g. `stream.realm` would throw an exception rather than doing a database query), since that would be a useful trick to have in our toolbox even if we don't use it here.

showell commented:
I'm definitely in favor of just doing the denormalization thing now.  Do you have a more specific suggestion on what the model should be, particularly with regard to Zephyr?

For preventing realm access without going the full StreamLite path, I did a bit of research but didn't come up with anything.  It's definitely not as simple as overriding `realm`; Django's descriptors prevent that.

StreamLite does have the additional benefit of being both smaller and quicker to deserialize.  If we can address the `is_public` nuisance, then the only ugly thing remaining is the id assignment.  Unfortunately, I don't have an elegant way to fix that.  (I mean, it works fine, it's just kind of inconsistent with the rest of the code.)

showell commented:
I renamed this PR to "wip", as the feedback here is to figure out a way to denormalize the Zephyr checks to simplify stream.is_public().  I also rebased it to reflect that some of the commits from the original PR are now on master.

showell commented:
This PR has had some stuff merged via other branches, and it has some conflicts, but I want to keep it open for a bit, as there a couple commits here that are worth cherry-picking or cribbing, and I may get pulled away from this.

timabbott commented:
Is this still active?

showell commented:
I haven't completely abandoned it yet, but I'm also not actively working on it.

timabbott commented:
Cool.  It'd be sorta helpful for my using TinglingGit if you rebased this to at least drop the commits that are already merged, since it shows up often as something conflicting with migrations we'd do, but I think actually most of that is in already-merged commits.

showell commented:
Yeah, I want to re-visit this branch eventually.  I think there's some good stuff here, but it's not urgent, and it requires a ton of focus. I rebased somewhat recently, but it obviously got stale again. :(
